flox/core.py

```diff
@@ -1584,18 +1596,33 @@ def _assert_by_is_aligned(shape: tuple[int, ...], by: T_Bys):
             )
 
 
+@overload
+def _convert_expected_groups_to_index(
+    expected_groups: tuple[None, ...], isbin: Sequence[bool], sort: bool
+) -> tuple[None, ...]:
+    ...
+
+
+@overload
 def _convert_expected_groups_to_index(
     expected_groups: T_ExpectTuple, isbin: Sequence[bool], sort: bool
 ) -> T_ExpectIndexTuple:
-    out: list[pd.Index | None] = []
+    ...
+
+
+def _convert_expected_groups_to_index(
+    expected_groups: T_ExpectOptTuple, isbin: Sequence[bool], sort: bool
+) -> T_ExpectIndexOptTuple:
+    out: list[T_ExpectIndexOpt] = []
     for ex, isbin_ in zip(expected_groups, isbin):
         if isinstance(ex, pd.IntervalIndex) or (isinstance(ex, pd.Index) and not isbin_):
+            e = ex
             if sort:
-                ex = ex.sort_values()
-            out.append(ex)
+                e = e.sort_values()
+            out.append(e)
```

Comment:
Can we revert this please?

Reply:
Nope, doesn't work:
```
flox/core.py:1621: error: Argument 1 to "append" of "list" has incompatible type "Sequence[Any] | ndarray[Any, Any] | Index | None"; expected "Index | None"  [arg-type]
```
Tested in [8e55d3a](https://github.com/xarray-contrib/flox/pull/235/commits/8e55d3a62181153bcd011d99fc73af5b3a01eb6b)

Reply:
Seems like mypy isn't narrowing typing. I'd prefer a `type: ignore` instead.

Reply:
Add reveal_types:
```python
    out: list[T_ExpectIndexOpt] = []
    for ex, isbin_ in zip(expected_groups, isbin):
        reveal_type(ex) #     flox/core.py:1618: note: Revealed type is "Union[typing.Sequence[Any], numpy.ndarray[Any, Any], pandas.core.indexes.base.Index, None]"
        if isinstance(ex, pd.IntervalIndex) or (isinstance(ex, pd.Index) and not isbin_):
            reveal_type(ex) # flox/core.py:1620: note: Revealed type is "Union[pandas.core.indexes.interval.IntervalIndex[Any], pandas.core.indexes.base._IntIndexType]"
            e = ex
            reveal_type(e) #  flox/core.py:1622: note: Revealed type is "Union[pandas.core.indexes.interval.IntervalIndex[Any], pandas.core.indexes.base._IntIndexType]"
            if sort:
                e = e.sort_values()
            reveal_type(e) #  flox/core.py:1625: note: Revealed type is "Union[pandas.core.indexes.interval.IntervalIndex[Any], pandas.core.indexes.base._IntIndexType]"

            out.append(e)
```

Reply:
It is the redefinition of ex that's the problem: `ex = ex.sort_values()`.
When redefining a variable mypy always takes the first seen typing of that variable, even though it has already been narrowed down.

At least mypy is consistent with this, it really doesn't like redefinitions with different typing.

I added an `else` without the redefinition instead.

Reply:
Thanks for tracking that down!